File: mycollab-dao/src/main/java/com/esofthead/mycollab/core/db/query/StringParam.java
package com.esofthead.mycollab.core.db.query;

import com.esofthead.mycollab.core.MyCollabException;
import com.esofthead.mycollab.core.arguments.NoValueSearchField;
import com.esofthead.mycollab.core.arguments.OneValueSearchField;
import com.esofthead.mycollab.core.arguments.SearchField;

/**
 * 
 * @author MyCollab Ltd.
 * @since 4.0
 * 
 */
public class StringParam extends ColumnParam {
	public static final String IS = "is";
	public static final String IS_NOT = "isn't";
	public static final String CONTAINS = "contains";
	public static final String NOT_CONTAINS = "doesn't contains";
	public static final String IS_EMPTY = "is empty";
	public static final String IS_NOT_EMPTY = "is not empty";

	private static String NULL_EXPR = "%s.%s is null";
	private static String NOT_NULL_EXPR = "%s.%s is not null";
	private static String EQUAL_EXPR = "%s.%s = ";
	private static String NOT_EQUAL_EXPR = "%s.%s <> ";
	private static String LIKE_EXPR = "%s.%s like ";
	private static String NOT_LIKE_EXPR = "%s.%s not like ";

	public static String[] OPTIONS = { IS, IS_NOT, CONTAINS, NOT_CONTAINS,
			IS_EMPTY, IS_NOT_EMPTY };

	public StringParam(String id, String displayName, String table,
			String column) {
		super(id, displayName, table, column);
	}

	public SearchField buildSearchField(String prefixOper, String compareOper,
			String value) {
		switch (compareOper) {
		case StringParam.IS_EMPTY:
			return this.buildStringParamIsNull(prefixOper);
		case StringParam.IS_NOT_EMPTY:
			return this.buildStringParamIsNotNull(prefixOper);
		case StringParam.IS:
			return this.buildStringParamIsEqual(prefixOper, value);
		case StringParam.IS_NOT:
			return this.buildStringParamIsNotEqual(prefixOper, value);
		case StringParam.CONTAINS:
			return this.buildStringParamIsLike(prefixOper, value);
		case StringParam.NOT_CONTAINS:
			return this.buildStringParamIsNotLike(prefixOper, value);
		default:
			throw new MyCollabException("Not support yet");
		}
	}

	public NoValueSearchField buildStringParamIsNull(String oper) {
		return new NoValueSearchField(oper, String.format(NULL_EXPR,
				this.getTable(), this.getColumn()));
	}

	public NoValueSearchField andStringParamIsNull() {
		return buildStringParamIsNull(SearchField.AND);
	}

	public NoValueSearchField orStringParamIsNull() {
		return buildStringParamIsNull(SearchField.OR);
	}

	public NoValueSearchField buildStringParamIsNotNull(String oper) {
		return new NoValueSearchField(oper, String.format(NOT_NULL_EXPR,
				this.getTable(), this.getColumn()));
	}

	public NoValueSearchField andStringParamIsNotNull() {
		return buildStringParamIsNotNull(SearchField.AND);
	}

	public NoValueSearchField orStringParamIsNotNull() {
		return buildStringParamIsNull(SearchField.OR);
	}

	public OneValueSearchField buildStringParamIsEqual(String oper, Object value) {
		return new OneValueSearchField(oper, String.format(EQUAL_EXPR,
				this.getTable(), this.getColumn()), value);
	}

	public OneValueSearchField andStringParamIsEqual(Object value) {
		return buildStringParamIsEqual(SearchField.AND, value);
	}

	public OneValueSearchField orStringParamIsEqual(Object value) {
		return buildStringParamIsEqual(SearchField.OR, value);
	}

	public OneValueSearchField buildStringParamIsNotEqual(String oper,
			Object value) {
		return new OneValueSearchField(oper, String.format(NOT_EQUAL_EXPR,
				this.getTable(), this.getColumn()), value);
	}

	public OneValueSearchField andStringParamIsNotEqual(Object value) {
		return buildStringParamIsNotEqual(SearchField.AND, value);
	}

	public OneValueSearchField orStringParamIsNotEqual(StringParam param,
			Object value) {
		return buildStringParamIsNotEqual(SearchField.OR, value);
	}

	public OneValueSearchField buildStringParamIsLike(String oper, Object value) {
		return new OneValueSearchField(oper, String.format(LIKE_EXPR,
				this.getTable(), this.getColumn()), "%" + value + "%");
	}

	public OneValueSearchField andStringParamIsLike(Object value) {
		return buildStringParamIsLike(SearchField.AND, value);
	}

	public OneValueSearchField orStringParamIsLike(Object value) {
		return buildStringParamIsLike(SearchField.OR, value);
	}

	public OneValueSearchField buildStringParamIsNotLike(String oper,
			Object value) {
		return new OneValueSearchField(oper, String.format(NOT_LIKE_EXPR,
				this.getTable(), this.getColumn()), "%" + value + "%");
	}

	public OneValueSearchField andStringParamIsNotLike(Object value) {
		return buildStringParamIsNotLike(SearchField.AND, value);
	}

	public OneValueSearchField orStringParamIsNotLike(Object value) {
		return buildStringParamIsNotLike(SearchField.OR, value);
	}
}

File: mycollab-dao/src/main/java/com/mycollab/db/query/StringParam.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.db.query

import com.mycollab.common.i18n.QueryI18nEnum.StringI18nEnum.*
import com.mycollab.core.MyCollabException
import com.mycollab.db.arguments.NoValueSearchField
import com.mycollab.db.arguments.OneValueSearchField
import com.mycollab.db.arguments.SearchField

/**
 * @author MyCollab Ltd.
 * @since 4.0
 */
class StringParam(id: String, table: String, column: String) : ColumnParam(id, table, column) {

    fun buildSearchField(prefixOper: String, compareOper: String, value: String): SearchField {
        val compareVal = valueOf(compareOper)
        return when (compareVal) {
            IS_EMPTY -> this.buildStringParamIsNull(prefixOper)
            IS_NOT_EMPTY -> this.buildStringParamIsNotNull(prefixOper)
            IS -> this.buildStringParamIsEqual(prefixOper, value)
            IS_NOT -> this.buildStringParamIsNotEqual(prefixOper, value)
            CONTAINS -> this.buildStringParamIsLike(prefixOper, value)
            NOT_CONTAINS -> this.buildStringParamIsNotLike(prefixOper, value)
            else -> throw MyCollabException("Not support yet")
        }
    }

    fun buildStringParamIsNull(oper: String): NoValueSearchField {
        val NULL_EXPR = "%s.%s is null"
        return NoValueSearchField(oper, String.format(NULL_EXPR,
                this.table, this.column))
    }

    fun andStringParamIsNull(): NoValueSearchField {
        return buildStringParamIsNull(SearchField.AND)
    }

    fun orStringParamIsNull(): NoValueSearchField {
        return buildStringParamIsNull(SearchField.OR)
    }

    fun buildStringParamIsNotNull(oper: String): NoValueSearchField {
        val NOT_NULL_EXPR = "%s.%s is not null"
        return NoValueSearchField(oper, String.format(NOT_NULL_EXPR, this.table, this.column))
    }

    fun andStringParamIsNotNull(): NoValueSearchField {
        return buildStringParamIsNotNull(SearchField.AND)
    }

    fun orStringParamIsNotNull(): NoValueSearchField {
        return buildStringParamIsNull(SearchField.OR)
    }

    fun buildStringParamIsEqual(oper: String, value: Any): OneValueSearchField {
        val EQUAL_EXPR = "%s.%s = "
        return OneValueSearchField(oper, String.format(EQUAL_EXPR, this.table, this.column), value)
    }

    fun andStringParamIsEqual(value: Any): OneValueSearchField {
        return buildStringParamIsEqual(SearchField.AND, value)
    }

    fun orStringParamIsEqual(value: Any): OneValueSearchField {
        return buildStringParamIsEqual(SearchField.OR, value)
    }

    fun buildStringParamIsNotEqual(oper: String, value: Any): OneValueSearchField {
        val NOT_EQUAL_EXPR = "%s.%s <> "
        return OneValueSearchField(oper, String.format(NOT_EQUAL_EXPR, this.table, this.column), value)
    }

    fun andStringParamIsNotEqual(value: Any): OneValueSearchField {
        return buildStringParamIsNotEqual(SearchField.AND, value)
    }

    fun orStringParamIsNotEqual(param: StringParam, value: Any): OneValueSearchField {
        return buildStringParamIsNotEqual(SearchField.OR, value)
    }

    fun buildStringParamIsLike(oper: String, value: Any): OneValueSearchField {
        val LIKE_EXPR = "%s.%s like "
        return OneValueSearchField(oper, String.format(LIKE_EXPR, this.table, this.column), "%$value%")
    }

    fun andStringParamIsLike(value: Any): OneValueSearchField {
        return buildStringParamIsLike(SearchField.AND, value)
    }

    fun orStringParamIsLike(value: Any): OneValueSearchField {
        return buildStringParamIsLike(SearchField.OR, value)
    }

    fun buildStringParamIsNotLike(oper: String, value: Any): OneValueSearchField {
        val NOT_LIKE_EXPR = "%s.%s not like "
        return OneValueSearchField(oper, String.format(NOT_LIKE_EXPR,
                this.table, this.column), "%$value%")
    }

    fun andStringParamIsNotLike(value: Any): OneValueSearchField {
        return buildStringParamIsNotLike(SearchField.AND, value)
    }

    fun orStringParamIsNotLike(value: Any): OneValueSearchField {
        return buildStringParamIsNotLike(SearchField.OR, value)
    }

    companion object {

        @JvmField val OPTIONS = arrayOf(IS, IS_NOT, CONTAINS, NOT_CONTAINS, IS_EMPTY, IS_NOT_EMPTY)
    }
}
